rfcs/2023-radix-form-primitive.md

```diff
@@ -0,0 +1,301 @@
+
+Forms are a huge part of building for the web, and whilst we do offer building blocks for it (things like `Checkbox`, `Select`, etc) we do not yet have an overarching solution for creating accessible forms. This RFC aims to solve that.
+
+Additionally, similarly to how Radix introduced a great out of the box experience for specific components (not having to wire state or refs for example) we see a big opportunity to do the same for forms as this is a space where we see a lot of people struggle or overcomplicate things.
+
+At [WorkOS](https://workos.com), we have also been working on components specifically geared towards building great authentication experiences (see [Radix Auth RFC](https://github.com/radix-ui/primitives/pull/1978)). As these rely heavily on forms, we thought it would make sense to provide some fundations in Radix to help with this.
+
+## Detailed design
+
+At the most basic level, the `Form` primitive is a regular Radix primitives package which expose a component split in multiple parts like other primitives. It offers a declarative way to use controls and labels, as well as a validation rules that are built on top of the native browser [constraint validation API](https://developer.mozilla.org/en-US/docs/Web/HTML/Constraint_validation).
+
+It abstracts away all of the complexity of accessibility and validation, using standard HTML attributes, as well as aria attributes.
+
+### API example
+
+Let's look at an example:
+
+```jsx
+import * as React from 'react';
+import * as Form from '@radix-ui/react-form';
+
+function Page() {
+  return (
+    <Form.Root
+      onSubmit={(event) => {
+        // `onSubmit` only triggered if it passes client-side validation
+        const data = Object.fromEntries(new FormData(event.currentTarget));
+        console.log(data);
+        event.preventDefault();
+      }}
+    >
+      <Form.Field name="name">
+        <Form.Label>Full name</Form.Label>
+        <Form.Control /> {/* renders a `input[type="text"]` by default */}
+        <Form.ClientMessage type="missingValue">Please enter your name.</Form.ClientMessage>
+      </Form.Field>
+
+      <Form.Field name="email">
+        <Form.Label>Email address</Form.Label>
+        <Form.Control type="email" />
+        <Form.ClientMessage type="missingValue">Please enter your email.</Form.ClientMessage>
+        <Form.ClientMessage type="typeMismatch">Please provide a valid email.</Form.ClientMessage>
+      </Form.Field>
+
+      <Form.Field name="age">
+        <Form.Label>Age</Form.Label>
+        <Form.Control type="number" min="18" max="57" step="1" />
+        <Form.ClientMessage type="missingValue">Please enter your age.</Form.ClientMessage>
+        <Form.ClientMessage type="rangeOverflow">You must be 18 or older.</Form.ClientMessage>
+        <Form.ClientMessage type="rangeUnderflow">You must be 57 or younger.</Form.ClientMessage>
+        <Form.ClientMessage type="stepMismatch">Please enter a whole number.</Form.ClientMessage>
```

Comment:
This can get a little confusing when there are several possible types of errors for an input. It may not be the best way to create a `Form.ClientMessage` for each `type`.

What about rendering just one `<Form.ClientMessage />` and giving it choices of types and error messages to render? Here is an example of what I believe may be a better approach (or not):

```jsx
const ageErrorsMessages = {
   missingValue: 'Please enter your age.',
   rangeOverflow: 'You must be 18 or older.',
   rangeUnderflow: 'You must be 57 or younger.',
   stepMismatch: 'Please enter a whole number.',
   customErrors: [
      {
         message: 'This is my first custom error message.',
         validate: (value, fields) => value === 'John'
      },
      // could do it forever
   ]
}

// ...

<Form.ClientMessage messages={ageErrorsMessages} />
```

If you like this approach, we could rename the `Form.ClientMessage` to something like `Form.ClientMessages` to handle multiple options.

Reply:
That makes sense! Thank you for clarifying all this stuff! Now, I agree that the current implementation is a more flexible and better approach.

Reply:
Yeah I think we're unlikely to change the 1 part = 1 node requirement, seeing as that's how Radix primitives work in general. Also it's kinda trivial to build abstractions around it.